corehq/util/doc_processor/couch.py

```diff
@@ -0,0 +1,135 @@
+            last_view_kwargs['startkey'] = [next_doc_type]
+            last_view_kwargs['endkey'] = [next_doc_type, {}]
+        return last_args, last_view_kwargs
+
+
+def ResumableDocsByTypeIterator(db, doc_types, iteration_key, chunk_size=100, view_event_handler=None):
+    """Perform one-time resumable iteration over documents by type
+
+    Iteration can be efficiently stopped and resumed. The iteration may
+    omit documents that are added after the iteration begins or resumes
+    and may include deleted documents.
+
+    :param db: Couchdb database.
+    :param doc_types: A list of doc type names to iterate on.
+    :param iteration_key: A unique key identifying the iteration. This
+    key will be used in combination with `doc_types` to maintain state
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param chunk_size: Number of documents to yield before updating the
+    iteration checkpoint. In the worst case about this many documents
+    that were previously yielded may be yielded again if the iteration
+    is stopped and later resumed.
+    """
+
+    def data_function(**view_kwargs):
+        return db.view('all_docs/by_doc_type', **view_kwargs)
+
+    doc_types = sorted(doc_types)
+    data_function.__name__ = " ".join(doc_types)
+
+    view_kwargs = {
+        'limit': chunk_size,
+        'include_docs': True,
+        'reduce': False,
+        'startkey': [doc_types[0]],
+        'endkey': [doc_types[0], {}]
+    }
+
+    args_provider = ResumableDocsByTypeArgsProvider(view_kwargs, doc_types)
+
+    class ResumableDocsIterator(ResumableFunctionIterator):
+        def __iter__(self):
+            for result in super(ResumableDocsIterator, self).__iter__():
+                yield result['doc']
+
+    def item_getter(doc_id):
+        try:
+            return {'doc': db.get(doc_id)}
+        except ResourceNotFound:
+            pass
+
+    return ResumableDocsIterator(iteration_key, data_function, args_provider, item_getter, view_event_handler)
+
+
+class CouchProcessorProgressLogger(ProcessorProgressLogger):
+    """
+    :param doc_types: List of doc_types that are being processed
+    """
+    def __init__(self, doc_types):
+        self.doc_types = [t[0] if isinstance(t, tuple) else t.__name__ for t in doc_types]
+
+    def progress_starting(self, total, previously_visited):
+        print("Processing {} documents{}: {}...".format(
+            total,
+            " (~{} already processed)".format(previously_visited) if previously_visited else "",
+            ", ".join(sorted(self.doc_types))
+        ))
+
+
+class CouchDocumentProvider(DocumentProvider):
+    """Document provider for couch documents.
+
+    All documents must live in the same couch database.
+
+    :param iteration_key: unique key to identify the document iterator
+    :param doc_types: Dict of `doc_type_name: model_class` pairs.
```

Comment:
Should say something like "An ordered sequence. Each item in the sequence should be either a doc type class or a tuple `(doc_type_name_string, doc_type_class)`"